File: leetcode_alg/solution223.cpp
//
// Created by wangheng on 1/7/24.
//

#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int computeArea(int ax1, int ay1, int ax2, int ay2, int bx1, int by1, int bx2, int by2) {

        // 判断是否相交
        if(bx1 >= ax2 || ax1 >= bx2 || by1 >= ay2 || ay1 >= by2){
            // 不相交, 返回两者面积总和
            return (ax2 - ax1) * (ay2 - ay1) + (bx2 - bx1) * (by2 - by1);
        }

        // 相交情况
        int cx1, cy1, cx2, cy2;
        cx1 = max(ax1, bx1);
        cy1 = max(ay1, by1);

        cx2 = min(ax2, bx2);
        cy2 = min(ay2, by2);

        int area = (cx2 - cx1) * (cy2 - cy1);
        return (ax2 - ax1) * (ay2 - ay1) + (bx2 - bx1) * (by2 - by1) - area;
    }
};

int main(int argc, char *argv[]) {

    int ax1, ay1, ax2, ay2, bx1, by1, bx2, by2;
    cin >> ax1 >> ay1 >> ax2 >> ay2 >> bx1 >> by1 >> bx2 >> by2;

    auto solution = new Solution;

    cout<<"结果:"<<solution->computeArea(ax1,ay1, ax2, ay2, bx1, by1, bx2, by2);

    delete solution;
    return 0;
}

File: golang/solution/solution223.go
package solution

import "fmt"

// 计算两个矩形覆盖的面积

func computeArea(ax1 int, ay1 int, ax2 int, ay2 int, bx1 int, by1 int, bx2 int, by2 int) int {
	if bx1 >= ax2 || bx2 <= ax1 || by1 >= ay2 || by2 <= ay1 {
		// no intersection
		w1, h1 := (ax2 - ax1), (ay2 - ay1)
		w2, h2 := (bx2 - bx1), (by2 - by1)
		return w1*h1 + w2*h2
	}

	var max func(int, int) int = func(i, j int) int {
		if i < j {
			return j
		}
		return i
	}

	var min func(int, int) int = func(i, j int) int {
		if i < j {
			return i
		}
		return j
	}

	cx1, cy1 := max(ax1, bx1), max(ay1, by1)
	cx2, cy2 := min(ax2, bx2), min(ay2, by2)

	crossArea := (cx2 - cx1) * (cy2 - cy1)

	w1, h1 := (ax2 - ax1), (ay2 - ay1)
	w2, h2 := (bx2 - bx1), (by2 - by1)
	return w1*h1 + w2*h2 - crossArea
}

func Solution223() {
	fmt.Println(computeArea(-3, 0, 3, 4, 0, -1, 9, 2))
}
